File: packages/sdk-node/src/sessions.ts
import type { HttpClient } from "./http.js";
import type { ExecOptions, ExecResult, ServiceId, Session } from "./types.js";
import { encodeExecOptions } from "./exec_options.js";

/**
 * Sessions bundle a set of phantom tokens that can be handed to a sandbox.
 * Each session's tokens die when the session is closed.
 */
export class Sessions {
  constructor(private readonly http: HttpClient) {}

  /** Create a new session. Returns the full record including the env map. */
  async create(params: {
    services: ServiceId[];
    ttlSecs?: number;
    /** Per-service allow-list of path globs (`*` only supported as trailing wildcard). */
    scopes?: Partial<Record<ServiceId, string[]>>;
  }): Promise<Session> {
    const body: {
      services: ServiceId[];
      ttl_secs?: number;
      scopes?: Partial<Record<ServiceId, string[]>>;
    } = { services: params.services };
    if (params.ttlSecs !== undefined) body.ttl_secs = params.ttlSecs;
    if (params.scopes !== undefined) body.scopes = params.scopes;
    return this.http.request<Session>({
      method: "POST",
      path: "/v1/sessions",
      body,
    });
  }

  /** List every session, newest first. */
  async list(): Promise<Session[]> {
    return this.http.request<Session[]>({
      method: "GET",
      path: "/v1/sessions",
    });
  }

  /** Fetch a session by id. */
  async get(id: string): Promise<Session> {
    return this.http.request<Session>({
      method: "GET",
      path: `/v1/sessions/${encodeURIComponent(id)}`,
    });
  }

  /** Execute a command inside this session's jail. */
  async exec(
    id: string,
    params: {
      cmd: string;
      args?: string[];
      timeoutSecs?: number;
      memoryMb?: number;
    } & ExecOptions,
  ): Promise<ExecResult> {
    const body: Record<string, unknown> = { cmd: params.cmd };
    if (params.args) body.args = params.args;
    if (params.timeoutSecs !== undefined) body.timeout_secs = params.timeoutSecs;
    if (params.memoryMb !== undefined)    body.memory_mb    = params.memoryMb;
    encodeExecOptions(body, params);
    return this.http.request<ExecResult>({
      method: "POST",
      path: `/v1/sessions/${encodeURIComponent(id)}/exec`,
      body,
    });
  }

  /** Close a session. Revokes every phantom token it issued. */
  async close(id: string): Promise<void> {
    await this.http.request<void>({
      method: "DELETE",
      path: `/v1/sessions/${encodeURIComponent(id)}`,
    });
  }
}

File: packages/sdk-python/src/agentjail/sessions.py
"""Sessions — bundle phantom tokens + exec in the jail."""

from __future__ import annotations

from typing import Any

from ._http import HttpClient
from .types import ExecResult, ServiceId, Session


class Sessions:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        *,
        services: list[ServiceId],
        ttl_secs: int | None = None,
        scopes: dict[ServiceId, list[str]] | None = None,
    ) -> Session:
        body: dict[str, Any] = {"services": services}
        if ttl_secs is not None:
            body["ttl_secs"] = ttl_secs
        if scopes is not None:
            body["scopes"] = scopes
        return self._http.request("POST", "/v1/sessions", json=body)

    def list(self) -> list[Session]:
        return self._http.request("GET", "/v1/sessions")

    def get(self, session_id: str) -> Session:
        return self._http.request("GET", f"/v1/sessions/{session_id}")

    def close(self, session_id: str) -> None:
        self._http.request("DELETE", f"/v1/sessions/{session_id}")

    def exec(
        self,
        session_id: str,
        *,
        cmd: str,
        args: list[str] | None = None,
        timeout_secs: int | None = None,
        memory_mb: int | None = None,
        network: Any = None,
        seccomp: str | None = None,
        cpu_percent: int | None = None,
        max_pids: int | None = None,
    ) -> ExecResult:
        body: dict[str, Any] = {"cmd": cmd}
        if args is not None:
            body["args"] = args
        if timeout_secs is not None:
            body["timeout_secs"] = timeout_secs
        if memory_mb is not None:
            body["memory_mb"] = memory_mb
        if network is not None:
            body["network"] = network
        if seccomp is not None:
            body["seccomp"] = seccomp
        if cpu_percent is not None:
            body["cpu_percent"] = cpu_percent
        if max_pids is not None:
            body["max_pids"] = max_pids
        return self._http.request("POST", f"/v1/sessions/{session_id}/exec", json=body)
